feat: add a ImmutableBuffer

kristian opened the pull request:
ImmutableBuffer throws a UnsupportedOperationException for anything that writes to the buffer. This disadvantage is compensated by the advantage the ImmutableBuffer can be transfered over the local event bus very efficiently without beeing copied.

pk-work commented on src/test/java/io/neonbee/test/base/NeonBeeTestBase.java:
```diff
@@ -113,26 +143,42 @@ public void setUp(Vertx vertx, VertxTestContext testContext, TestInfo testInfo)
-                    ServerVerticle dummyServerVerticle = createDummyServerVerticle(testInfo);
+                    try {
+                        // just to be sure, try to fetch a new port to use
+                        options.setServerPort(SystemHelper.getFreePort());
```
The downside of this is, that the ServerVerticleConfig  in /config folder is not longer in sync with the instantiated object .. Maybe we should write the "new" config also back into the /config foler.

kristian replied:
Yes I knew. I tolerated this, because this is only covering an edge case, which should normally not ocour. However all methods which read the port, anyways read it from the options. The config is just there "in addition". So we should be good with this.

kristian commented on src/test/java/io/neonbee/NeonBeeExtension.java:
```diff
@@ -107,25 +108,56 @@ private Object unpack(Object object) {
+        }
+    }
+
+    public static NeonBeeOptions.Mutable defaultOptions() {
```
I moved it into an own "OptionsHelper" class

kristian commented on src/test/java/io/neonbee/job/JobVerticleTest.java:
```diff
@@ -93,6 +98,12 @@ public String deploymentID() {
+    @Override
+    protected void adaptOptions(TestInfo testInfo, NeonBeeOptions.Mutable options) {
+        options.setActiveProfiles(List.of(ALL, NO_WEB));
+        options.setDisableJobScheduling(false);
```
Because WatchVerticle is no JobVerticle :)

pk-work replied:
LoL good point! Maybe it should become one in the future.

kristian commented on src/test/java/io/neonbee/test/helper/SystemHelper.java:
```diff
@@ -2,19 +2,48 @@
 
     /**
-     * This method tries to find a free port on the system and returns it.
+     * This method tries to find a free ephemeral port on the system and returns it. This method will never return the
```
But it does pick a random (?) port, the new implementation will guarantee, that for one test run (in one JVM) no same port is returned twice.

kristian commented on src/test/java/io/neonbee/test/listeners/StaleVertxChecker.java:
```diff
@@ -0,0 +1,159 @@
+        }
+
+        // try to find instances associated to this test or any instances that we cannot associate to any test
+        Set<Vertx> unassociatedVertxInstances = new HashSet<>(), associatedVertxInstances = new HashSet<>(), // NOPMD
+                notAssociatedVertxInstances = new HashSet<>();
```
No, this does not make sense to me. Then we should adapt the PMD ruling.

If I want to define multiple sets with the same generic type, which MUST be of the same generic type, it doesn't make sense to declare it multiple times, which is the bigger code smell to me, because it assumes that the sets can be of different generic type, which is not true:

Set<Vertx> unassociatedVertxInstances = new HashSet<>();

Set<Vertx> associatedVertxInstances = new HashSet<>();

Set<Vertx> notAssociatedVertxInstances = new HashSet<>();

Thus it is in one declaration, because it makes it clear: "These sets MUST be compatible to each other".

Please let me know if the PMD rule should be adapted, or if you are fine with the NOPMD here.

kristian commented on build.gradle:
```diff
@@ -9,16 +9,16 @@ plugins {
+    id 'se.bjurr.violations.violations-gradle-plugin' version '1.50.16'
+    id 'com.github.johnrengelman.shadow' version '7.0.0'
+    id 'org.sonarqube' version '3.3'
+    //id 'team.yi.semantic-gitlog' version '0.5.17' // isn't compatible with Gradle 7.2 yet
```
At some point in time I updated everything, because I did assume it is a bug in the parallel execution of JUnit which was quite new. As JUnit parallel execution is build into Gradle I had to update gradle.

The plugin is only used for generating the changelog as far as I can tell. Executing the plugin with a lower Gradle is much less effort, than doing the whole Gradle update again. Thus I would argue we update it, as soon as it is available for Gradle 7. If the plugin will not be updated, we anyways have to check for a different plugin, as Gradle 8 is already around the corner and the Gradle 7 update would have been due anyways.

pk-work replied:
I agree, haven't thought about running the changelog task with an older version. This makes sense.but how can we achieve this?

kristian replied:
simply change the wrapper version and run the task, just work just fine :)

kristian commented on docs/internal/howto_debug_github_actions.md:
````diff
@@ -0,0 +1,37 @@
+  uses: mxschmitt/action-tmate@v3
+```
+
+It causes a [tmate](https://tmate.io/) session to be spun up (attention it is a public session, anybody could hook into it, when they monitor our build logs, protection would be possible with options of the `action-tmate`, however for an open source project like ours it is kind of irrelevant). The action *will wait* until you connect to the session via SSH and create a `continue` file (e.g. by us ...
````
Which credentials do you mean?

pk-work replied:
I mean credentials like this: https://github.com/SAP/neonbee/blob/8fce18285048a1c560af60aa053f766353ec775e/.github/workflows/voter.yml#L55

kristian replied:
Mhm, I don't know... let's be sure and always remove tmate again if debugging is done

I will provide a comment in the docs

kristian commented:
Thanks for the review @pk-work. I have three open discussions, would you mind providing your feedback? Thanks.

s4heid reviewed (changes requested):
Congrats, @kristian for this valuable change 🚀  there are only a few small things I would like to adjust.

kristian commented on src/main/java/io/neonbee/internal/verticle/WatchVerticle.java:
```diff
@@ -150,12 +152,19 @@ public WatchVerticle(Path watchPath, long interval, TimeUnit unit, boolean paral
         }
+
         (handleExisting ? handleExistingFiles(watchPath) : registerWatchKey(watchPath)).compose(
                 v -> Future.<Counter>future(promise -> getVertx().sharedData().getLocalCounter(counterName, promise)))
```
`getVertx()` of `AbstractVerticle` will internally return the reference to this.vertx, however in abstract classes, or classes that extension is intended (e.g. WatchVerticle) it is considered a code smell to use the direct reference to a field, if a getter is available. This enables for instance overriding getVertx in tests to return a mocked Vert.x instance instead.

kristian commented on src/main/java/io/neonbee/internal/verticle/WatchVerticle.java:
```diff
@@ -150,12 +152,19 @@ public WatchVerticle(Path watchPath, long interval, TimeUnit unit, boolean paral
     public void start(Promise<Void> startPromise) {
+        Vertx vertx = getVertx();
+        if (NeonBee.get(vertx).getOptions().doNotWatchFiles()) {
+            startPromise.fail("Should not watch files");
```
You are right, I changed it to a similar approach that we had in JobVerticle and undeploy the verticle immediately after start.

kristian commented on src/test/java/io/neonbee/NeonBeeInstanceConfiguration.java:
```diff
@@ -32,6 +32,8 @@
 
     boolean ignoreClassPath() default true;
 
+    boolean doNotWatchFiles() default true;
```
this class is only used in / for tests. therefore it is also in src/test/.... In non-tests the default is actually false, as you would expect.

kristian commented:
I added all the changes requested. @s4heid, @pk-work, I just added one more `feat:` as the previous to last commit. Could you please do a code review specifically for this one? The rest is changed as you reviewed it already. Afterwards we should be able to finally merge this beast! :)

pk-work commented on src/main/java/io/neonbee/internal/verticle/WatchVerticle.java:
```diff
@@ -154,7 +156,13 @@ public WatchVerticle(Path watchPath, long interval, TimeUnit unit, boolean paral
     public void start(Promise<Void> startPromise) {
         Vertx vertx = getVertx();
         if (NeonBee.get(vertx).getOptions().doNotWatchFiles()) {
```
In my opionion the we shouldn't check in WatchVerticle if the NeonBee option "doNotWatchFiles" ist set or not. It is a NeonBeeOption and should only change the behavior of NeonBee. If I want to deploy a Module, which contains a WatchVerticle I don't want that this NeonBee option prevents it.

We have 2 verticles that extends the WatchVerticle DeployerVerticle and ModelRefresher. If NeonBee option "doNotWatchFiles" is set, we simply shouldn't deploy this verticles.

kristian replied:
I see your point, however then we have a different behaviour in `JobVerticle` than we have in `WatchVerticle`. And if you don't want to have _any_ files watched, that's what the NeonBee option is for. The question is: Should we "enforce" it, then I would say, adding it to the `WatchVerticle` is kind of how we did it in the `JobVerticle` as well, or should we _not_ enforce it, then I would agree to your option. What do you think?

pk-work replied:
Maybe its the name of the option. I thought we introduced the option "doNotWatchFiles" with the idea to stop NeonBee watching the folders "models" and "verticles". Or did we had another idea in mind?

If this was the idea, I wouldn't enforce it. I got your point with the similarity to "shouldDisableJobScheduling", but then maybe we simply should rename the option into "doNotWatchModules" and try to "remove" the models folder as soon as possible. Or rename it to "doNotWatchResources" or "doNotWatchModelsAndModules".

To be honest in my opinion "shouldDisableJobScheduling" is also not really required. I'd prefer disabling the "unwanted" JobVerticles via their config.

kristian replied:
Well both options are particularly useful in tests, because in many / most tests you don't want to have files watch nor you want job verticles, Thus the default option for tests has been set to "true / true" so no jobs are started and no files will be watched. Maybe we should move this debate to an issue instead to have it in our backlog and resolve this conversation?

pk-work replied:
I don't know, because this change then introduces a behavior which maybe is unwanted and needs to be removed afterwards. On the other hand this PR is so big in the meanwhile, that it is hard to review. So maybe it makes sense to merge it now as it is and do the "removal" maybe later.

I like none of these options, so you can decide :)

kristian replied:
I would like to merge this change *now* :D if you permit! 👯

Then this beast and the whole chapter on non-running tests is hopefully and FINALLY closed!

kristian commented:
The change is now DONE. Can we merge, please? 👍

pk-work reviewed (approved):
LGTM To be honest I haven't had yet a final look again, but I think there were only small fixes.

kristian commented:
@pk-work, @s4heid I will click the button now! any further changes will go into a n ew PR!! 👍  THIS ONE IS DONE AND OVER!! Final agreement?

kristian commented:
It's done! 👍 Thanks a lot to both of you! @gedack, feel free to rebase your other pull requests now, which should fix the voter for them finally! *cheer*
